fedimint-api/src/task.rs

```diff
@@ -1,23 +1,132 @@
 use std::future::Future;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering::SeqCst;
+use std::sync::Arc;
 use std::time::{Duration, Instant};
 
+use anyhow::anyhow;
+use futures::lock::Mutex;
 use thiserror::Error;
 
 #[derive(Debug, Error)]
 #[error("deadline has elapsed")]
 pub struct Elapsed;
 
+#[derive(Debug, Default)]
+struct TaskGroupInner {
+    /// Was the shutdown requested, either externally or due to any task failure?
+    is_shutting_down: AtomicBool,
+    join: Mutex<Vec<(String, JoinHandle<()>)>>,
+}
+
+#[derive(Debug)]
+/// A group of task working together
+///
+/// Using this struct it is possible to spawn one or more
+/// main thread collabarating, which can cooperatively gracefully
+/// shut down, either due to external request, or failure of
+/// one of them.
+///
+/// Each thread should periodically check [`ThreadGroupHandle`] or rely
+/// on condition like channel disconnection
+#[derive(Clone, Default)]
+pub struct TaskGroup {
```

Comment:
I would have to polish this a bit before landing, but that's the idea - instead of spawning thread willy-nilly, have a primitive that keeps track of them, and allows to signal shutdown and wait for all of them to wrap up before exiting.